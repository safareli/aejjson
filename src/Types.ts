import { pipe } from "fp-ts/lib/function";
import * as D from "io-ts/Decoder";

export type Vessel = D.TypeOf<typeof Vessel>;
export type Port = D.TypeOf<typeof Port>;
export type PortCall = D.TypeOf<typeof PortCall>;
export type Schedule = D.TypeOf<typeof Schedule>;
export type LogEntryDeparture = D.TypeOf<typeof LogEntryDeparture>;
export type LogEntryArrival = D.TypeOf<typeof LogEntryArrival>;
export type LogEntryIsOmitted = D.TypeOf<typeof LogEntryIsOmitted>;
export type LogEntry = D.TypeOf<typeof LogEntry>;

export const Vessel = D.type({
  name: D.string,
  imo: D.number,
});

export const Vessels = D.array(Vessel);

export const Port = D.type({ id: D.string, name: D.string });

export const DateFromISOString: D.Decoder<string, Date> = {
  decode: (s: string) => {
    const d = new Date(s);
    return isNaN(d.getTime()) ? D.failure(s, "ISODateString") : D.success(d);
  },
};

export const DateTime = pipe(D.string, D.compose(DateFromISOString));

export const LogEntryDeparture = D.type({
  updatedField: D.literal("departure"),
  departure: DateTime,
  createdDate: DateTime,
});

export const LogEntryArrival = D.type({
  updatedField: D.literal("arrival"),
  arrival: DateTime,
  createdDate: DateTime,
});

export const LogEntryIsOmitted = D.type({
  updatedField: D.literal("isOmitted"),
  isOmitted: D.boolean,
  createdDate: DateTime,
});

export const LogEntry = D.sum("updatedField")({
  departure: LogEntryDeparture,
  arrival: LogEntryArrival,
  isOmitted: LogEntryIsOmitted,
});

export const PortCall = D.type({
  arrival: DateTime,
  departure: DateTime,
  createdDate: DateTime,
  isOmitted: D.boolean,
  service: D.string,
  port: Port,
  logEntries: D.array(LogEntry),
});

export const Schedule = D.type({
  vessel: Vessel,
  portCalls: D.array(PortCall),
});
