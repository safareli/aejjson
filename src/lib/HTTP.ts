import { Either } from "fp-ts/lib/Either";
import * as D from "io-ts/Decoder";

export type GET = typeof get;
export const get = async <T>(
  url: string,
  decoder: D.Decoder<unknown, T>
): Promise<Either<D.DecodeError, T>> => {
  return window
    .fetch(url)
    .then((res) => res.json())
    .then((json) => decoder.decode(json));
};
