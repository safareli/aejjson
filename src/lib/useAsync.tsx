import { useEffect, useState } from "react";

export type Async<T> =
  | { type: "loading" }
  | { type: "failed"; error: Error }
  | { type: "loaded"; result: T };

export function useAsync<T>(fn: () => Promise<T>, deps: any[]) {
  const [res, setRes] = useState<Async<T>>({ type: "loading" });
  useEffect(() => {
    let cancel = false;
    fn().then(
      (result) => {
        if (cancel) return;
        setRes({ type: "loaded", result });
      },
      (error) => {
        if (cancel) return;
        setRes({ type: "failed", error });
      }
    );
    return () => {
      cancel = true;
    };
  }, deps);
  return res;
}
