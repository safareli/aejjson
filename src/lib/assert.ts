import { Either, isLeft } from "fp-ts/lib/Either";

export const assertRight = <E, V>(
  value: Either<E, V>,
  msg?: string | ((error: E) => string)
): V => {
  if (isLeft(value)) {
    const message =
      msg == null ? "" : typeof msg === "string" ? msg : msg(value.left);
    console.log(message);
    throw new Error("Assertion failed; " + message);
  }
  return value.right;
};
