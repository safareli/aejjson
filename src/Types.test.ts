import { draw } from "io-ts/lib/Decoder";
import { assertRight } from "./lib/assert";
import { PortCall, Vessel } from "./Types";

it("decode", () => {
  const vessel: unknown = { imo: 9303807, name: "ABIDJAN EXPRESS" };
  const portCall: unknown = {
    arrival: "2018-12-28T23:12:00+00:00",
    departure: "2019-01-05T07:03:00+00:00",
    createdDate: "2019-01-16T09:04:49.048105+00:00",
    isOmitted: false,
    service: "Pacific South Loop 1",
    port: {
      id: "USLAX",
      name: "Los Angeles",
    },
    logEntries: [
      {
        updatedField: "arrival",
        arrival: "2018-12-28T23:12:00+00:00",
        departure: null,
        isOmitted: null,
        createdDate: "2019-01-16T09:04:49.048105+00:00",
      },
      {
        updatedField: "departure",
        arrival: null,
        departure: "2019-01-05T07:03:00+00:00",
        isOmitted: null,
        createdDate: "2019-01-16T09:04:49.048105+00:00",
      },
    ],
  };

  const vesselDecoded = assertRight(Vessel.decode(vessel), draw);
  const portCallDecoded = assertRight(PortCall.decode(portCall), draw);
  expect(vesselDecoded).toMatchSnapshot();
  expect(portCallDecoded).toMatchSnapshot();
});
