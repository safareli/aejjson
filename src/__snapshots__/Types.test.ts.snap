// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decode 1`] = `
Object {
  "imo": 9303807,
  "name": "ABIDJAN EXPRESS",
}
`;

exports[`decode 2`] = `
Object {
  "arrival": 2018-12-28T23:12:00.000Z,
  "createdDate": 2019-01-16T09:04:49.048Z,
  "departure": 2019-01-05T07:03:00.000Z,
  "isOmitted": false,
  "logEntries": Array [
    Object {
      "arrival": 2018-12-28T23:12:00.000Z,
      "createdDate": 2019-01-16T09:04:49.048Z,
      "updatedField": "arrival",
    },
    Object {
      "createdDate": 2019-01-16T09:04:49.048Z,
      "departure": 2019-01-05T07:03:00.000Z,
      "updatedField": "departure",
    },
  ],
  "port": Object {
    "id": "USLAX",
    "name": "Los Angeles",
  },
  "service": "Pacific South Loop 1",
}
`;
